virttest/libvirt_xml/devices/disk.py

```diff
@@ -240,6 +240,30 @@ def new_drivermetadata(self, **dargs):
             setattr(new_one, key, value)
         return new_one
 
+    def get_all_backingstore(self):
```

Comment:
If this is used to satisfy the requirement only from https://github.com/autotest/tp-libvirt/pull/3384. 
It sound likes below logic appear more primitive, and we may not need introduce so many code changes.
```
>>> import xml.etree.ElementTree as ET
>>> tree = ET.parse(disk.xml)
>>> backingstore_list = []
>>> root = tree.getroot()
>>> for elem in root.iter('backingStore'):
...     backingstore_list.append(elem)
    
```

Reply:
> If this is used to satisfy the requirement only from [autotest/tp-libvirt#3384](https://github.com/autotest/tp-libvirt/pull/3384).

No, it's not **only** for that pr, all backing chain related checking could use this implementation.
> It sound likes below logic appear more primitive, and we may not need introduce so many code changes.
> 
> ```
> >>> import xml.etree.ElementTree as ET
> >>> tree = ET.parse(disk.xml)
> >>> backingstore_list = []
> >>> root = tree.getroot()
> >>> for elem in root.iter('backingStore'):
> ...     backingstore_list.append(elem)
>     
> ```
I really don't think so. If we're all using primitive ways to parse xml, why should we implement all these xml classed, vmxml, all the devices: interface, disk...etc.

Reply:
@kylazhang 
@dzhengfy 
Could you also contribute to your thoughts?

Reply:
@kylazhang 
@dzhengfy 
Could you contribute to your thoughts ?

Reply:
Disagreed. Let's wait for other opinions.

Reply:
@chunfuwen Since you clearly think implementing get_all_backingstore() with your way is better, then how about you go ahead and **create another pr** with the implementation? Then we'll have the chance to **compare the two ways**, take opinions. I think it would be much clearer than just saying the words, right?

Reply:
@chloerh 
reference PR is available :https://github.com/avocado-framework/avocado-vt/pull/3062

Reply:
@chloerh @chunfuwen , 
My opinion:
1. I respect everyone's work for enhancing our projects. So as long as it has any improvement, I would like to accept it.
2. I understand different people have different minds, so we do not need single way to solve single problem. 
3. Please help to use different function name to make difference. @chunfuwen 
So I will approve this one because I do not have other questions for the codes themselves.

Reply:
@chunfuwen @chloerh Are you sure you are talking the same usage?
From my test, @chunfuwen 's object in list have the function like this
```
>>> dir(backingstore_list[0])
['__class__', '__copy__', '__deepcopy__', '__delattr__', '__delitem__', '__dir__', '__doc__', '__eq__', '__format__', '__ge__', '__getattribute__', '__getitem__', '__getstate__', '__gt__', '__hash__', '__init__', '__init_subclass__', '__le__', '__len__', '__lt__', '__ne__', '__new__', '__reduce__', '__reduce_ex__', '__repr__', '__setattr__', '__setitem__', '__setstate__', '__sizeof__', '__str__', '__subclasshook__', 'append', 'attrib', 'clear', 'extend', 'find', 'findall', 'findtext', 'get', 'getchildren', 'getiterator', 'insert', 'items', 'iter', 'iterfind', 'itertext', 'keys', 'makeelement', 'remove', 'set', 'tag', 'tail', 'text']
```
But the @chloerh 's elem.backingstore object is
```
>>> dir(bs)
['INITIALIZED', 'Source', '___all_slots__', '__all_slots__', '__canhaz__', '__class__', '__contains__', '__delattr__', '__delitem__', '__dict__', '__dict_del__', '__dict_get__', '__dict_set__', '__dir__', '__doc__', '__eq__', '__format__', '__ge__', '__get__', '__getattr__', '__getattribute__', '__getitem__', '__gt__', '__hash__', '__init__', '__init_subclass__', '__iter__', '__le__', '__len__', '__lt__', '__module__', '__ne__', '__new__', '__reduce__', '__reduce_ex__', '__repr__', '__schema_name__', '__set__', '__setattr__', '__setitem__', '__sizeof__', '__slots__', '__str__', '__subclasshook__', '__super_del__', '__super_get__', '__super_set__', '__uncompareable__', '__weakref__', 'backingstore', 'clear', 'copy', 'del_backingstore', 'del_format', 'del_index', 'del_source', 'del_type', 'del_validates', 'del_xmltreefile', 'format', 'fromkeys', 'get', 'get_backingstore', 'get_format', 'get_index', 'get_section_string', 'get_source', 'get_type', 'get_validates', 'get_xml', 'get_xmltreefile', 'index', 'items', 'keys', 'new_source', 'pop', 'popitem', 'restore', 'set_backingstore', 'set_format', 'set_index', 'set_source', 'set_type', 'set_validates', 'set_virsh', 'set_xml', 'set_xmltreefile', 'setdefault', 'source', 'type', 'update', 'validates', 'values', 'virsh', 'virt_xml_validate', 'xml', 'xmltreefile']
```
Please make sure the usage for this object list then we can discuss if these 2 methods can replace each other.

Reply:
@kylazhang they're not the same, but similar.  The purpose of this change is to make it simpler to access xml attrs/elements, just like what we did all along, like what we did with all those xml classes: vmxml, interface, disk, controller

The difference between these 2 functions:

- My function will return a list of backingstore type of objects, just like other xml objects: interface, controller, which could allow us to access with pre-defined attributes.
- @chunfuwen 's function will return a list of xml elements. Accessing values needs to parse xml tree manually.

Reply:
@kylazhang 
@hahakiki2010 

Go gack to orginal purpose of parsing source_file from backstore
For simplicity, assume that disk is available with get_backingstore
Below give more details how they works.
***********************************************************************
```
solution 1: https://github.com/avocado-framework/avocado-vt/pull/3062/ 

test_disk = Disk()
backstore_lst = test_disk.get_backingstore_list()
parse_source_file_list = [elem.find('source').get('file') or elem.find('source').get('name') for elem in backstore_lst]
print(parse_source_file_list)
['/home/storage/d/../c/c.img', '/home/storage/d/../c/../b/b.img', '/home/storage/d/../c/../b/../a/a.img', 'iqn.2021-05.com.virttest:emulated-iscsi.target/0']

*************************************************************************
solution 2 :

https://github.com/avocado-framework/avocado-vt/pull/3038

test_disk = Disk()
backstore_lst = test_disk.get_all_backingstore()
parse_source_file_lists = [bs.source.attrs.get('file') or bs.source.attrs.get('name') for bs in backstore_lst]
print(parse_source_file_lists)

['/home/storage/d/../c/c.img', '/home/storage/d/../c/../b/b.img', '/home/storage/d/../c/../b/../a/a.img', 'iqn.2021-05.com.virttest:emulated-iscsi.target/0']

```

Reply:
@chunfuwen I'm quite confused about what this comment is about.

If you'd like to get attribute `index`
```
>>> [bs.index for bs in bs_list]
['2', '3', '4', '5']
```
BTW, this example has been posted https://github.com/avocado-framework/avocado-vt/pull/3038#issuecomment-834184100  for 2 weeks, in case you haven't noticed.